csrc/optimization/layout_inference.cpp

```diff
@@ -0,0 +1,236 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/all_nodes.h>
+#include <ir/utils.h>
+#include <optimization/layout_inference.h>
+
+namespace nvfuser {
+
+namespace {
+
+class MemoryFormatInferencer : public OptOutConstDispatch {
+ public:
+  MemoryFormatInferencer(
+      std::unordered_map<const TensorView*, MemoryFormat>& format_map)
+      : format_map_(format_map) {}
+
+ private:
+  void handle(const UnaryOp*) override;
+  void handle(const BinaryOp*) override;
+  void handle(const BroadcastOp*) override;
+  // TODO: Add more propagation rules
+  //   void handle(const Reduction*) override;
+  //   void handle(const LoadStoreOp*) override;
+  //   void handle(const SqueezeOp*) override;
+  //   void handle(const ExpandOp*) override;
+
+ private:
+  // format_map_ records the stride order (memory format) of each TensorView.
+  // Since it only handles permutation from a rfactor domain to allocation
+  // domain, it can be interpreted as:
+  //
+  // e.g. TV0 rfactor domain [i0, i1, i2]
+  //            alloc domain [i0, i2, i1]
+  //           memory format   0,  2,  1
+  std::unordered_map<const TensorView*, MemoryFormat>& format_map_;
+};
+
+// UnaryOp propagation forward memory format from input to output
+void MemoryFormatInferencer::handle(const UnaryOp* op) {
+  TensorView* out = dynamic_cast<TensorView*>(op->out());
+  if (out == nullptr) {
+    return;
+  }
+  TensorView* in = op->in()->as<TensorView>();
+  if (auto iter = format_map_.find(in); iter != format_map_.end()) {
+    format_map_[out] = iter->second;
+  }
+}
+
+// BinaryOp propagation tries to merge the memory format of both inputs
+//
+//   1. when there's only one operand has a recorded memory format, it forwards
+//   that.
+//   2. When both tensor have recorded memory format. It breaks tie based on the
+//   innermost dimension. whichever operand has a "better match" dominates the
+//   output format, where a "better match" meaning "less broadcast dimensions on
+//   the inner dimension"
+//
+// e.g.
+//   lhs TV0 rfactor_dom [i0, i1, b2]
+//                         0   2   1
+//   rhs TV0 rfactor_dom [i3, i4, b5]
+//                         0   1   2
+//   if we go from innermost to outermost order:
+//       TV0 has i1 -> b2 -> i0
+//       TV1 has b5 -> i4 -> i3
+//   we see that TV0 encounters a non-broadcast iter domain first, so TV0 is the
+//   dominating tensor. We'll produce an output with stride order identical to
```

Comment:
> so TV0 is the dominating tensor

Why are we in favor of the memory format that first hits a non-broadcast? (I suspect it's something about vectorization, but the comment wasn't clear about that)

Reply:
I have the same question. Why is this better than just use `lhs`? @jjsjann123 Could you add the explanation to the code here as comment?

Reply:
I somehow feel that, what we should do is: if this binary op contains a broadcast concretization, then respect the one with most number of concrete IDs, otherwise, just use lhs. cc @naoyam

Reply:
If we are propagating only in the forward direction, it seems to me that we can't really know what will be the most advantageous stride order. For example if we later do a sum on some outer dimension then it might wind up that we would have preferred that dimension to be allocated inner-most, but we would need to propagate that information backwards. If we're sticking with forward-only, why not just use the first input's stride order for the output and call it a day? If we want to chase more optimality we could consider doing an iterative optimization on the segmented fusion, allowing the schedulers to specify weighted preferences for the allocation orderings of their inputs and propagating changes to the outputs using simple rules like the one here, but that optimization is a bigger change to tackle.

Reply:
> Why are we in favor of the memory format that first hits a non-broadcast? (I suspect it's something about vectorization, but the comment wasn't clear about that)

This one should have been updated. I was doing this earlier when I use a different propagation rule for broadcast, so I needed this trick to propagate nhwc.

```
tv0 = [i0 i1 i2 i3] @ {0 2 3 1}
bias0 = [i4] @ {0} -> broadcast_bias0 [b5 i4 b6 b7] @ {0 1 2 3}
```
But now I feel @zasdfgbnm's suggestion makes a lot more sense instead.

> I somehow feel that, what we should do is: if this binary op contains a broadcast concretization, then respect the one with most number of concrete IDs, otherwise, just use lhs.

This makes a lot more sense to me. i.e. favoring larger tensor (hopefully more concrete IDs would lead to a larger tensor). I'll update.